File: Data Structures & Algorithms/partition-equal-subset-sum/submission-6.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {boolean}
     */
    canPartition(nums) { 
        const sum = nums.reduce((acc, cur) => acc + cur, 0) 

        if (sum % 2 !== 0) return false 

        return this.dfs(0, 0, nums, sum / 2)
    } 

    dfs(i, cur, nums, target) { 
        if (cur === target) return true 
        if (cur > target || i === nums.length) return false 

        const res = this.dfs(i + 1, cur + nums[i], nums, target) || 
                    this.dfs(i + 1, cur, nums, target) 

        return res
    }
}

File: Data Structures & Algorithms/permutations/submission-6.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        added = set() 
        res = [] 

        def dfs(start, cur): 
            if len(cur) == len(nums): 
                res.append(cur.copy()) 
                return 

            for i in range(len(nums)): 
                curNum = nums[i] 
                if curNum in added: 
                    continue 
                added.add(curNum) 
                cur.append(curNum) 
                dfs(i + 1, cur) 
                added.discard(curNum) 
                cur.pop() 

        dfs(0, []) 
        return res